MemberToMemberDiscoveryTest

**com.hazelcast.spi.discovery.multicast.MemberToMemberDiscoveryTest.formClusterWithTwoMembersTest**
```java
java.lang.AssertionError: Cluster size is not correct. Expected: 2 Actual: 1
	at org.junit.Assert.fail(Assert.java:88)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSize(HazelcastTestSupport.java:915)
	at com.hazelcast.test.HazelcastTestSupport$13.run(HazelcastTestSupport.java:937)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1086)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:933)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:927)
	at com.hazelcast.spi.discovery.multicast.MemberToMemberDiscoveryTest.formClusterWithTwoMembersTest(MemberToMemberDiscoveryTest.java:58)
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/10127/testReport/junit/com.hazelcast.spi.discovery.multicast/MemberToMemberDiscoveryTest/formClusterWithTwoMembersTest/

**com.hazelcast.spi.discovery.multicast.MemberToMemberDiscoveryTest.formClusterWithTwoMembersTest**
```java
java.lang.AssertionError: Cluster size is not correct. Expected: 2 Actual: 1
	at org.junit.Assert.fail(Assert.java:88)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSize(HazelcastTestSupport.java:915)
	at com.hazelcast.test.HazelcastTestSupport$13.run(HazelcastTestSupport.java:937)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1086)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:933)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:927)
	at com.hazelcast.spi.discovery.multicast.MemberToMemberDiscoveryTest.formClusterWithTwoMembersTest(MemberToMemberDiscoveryTest.java:58)
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/10501/testReport/junit/com.hazelcast.spi.discovery.multicast/MemberToMemberDiscoveryTest/formClusterWithTwoMembersTest/

--- a/hazelcast/src/main/java/com/hazelcast/cluster/impl/TcpIpJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/cluster/impl/TcpIpJoiner.java
@@ -127,7 +127,7 @@ public class TcpIpJoiner extends AbstractJoiner {
     private void joinViaPossibleMembers() {
         try {
             blacklistedAddresses.clear();
-            Collection<Address> possibleAddresses = getPossibleAddresses();
+            Collection<Address> possibleAddresses = getPossibleAddressesForInitialJoin();
 
             boolean foundConnection = tryInitialConnection(possibleAddresses);
             if (!foundConnection) {
@@ -179,6 +179,10 @@ public class TcpIpJoiner extends AbstractJoiner {
         } catch (Throwable t) {
             logger.severe(t);
         }
+    }
+
+    protected Collection<Address> getPossibleAddressesForInitialJoin() {
+        return getPossibleAddresses();
     }
 
     @SuppressWarnings("checkstyle:npathcomplexity")

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/DiscoveryJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/DiscoveryJoiner.java
@@ -22,13 +22,17 @@ import com.hazelcast.instance.Node;
 import com.hazelcast.nio.Address;
 import com.hazelcast.spi.discovery.DiscoveryNode;
 import com.hazelcast.spi.discovery.integration.DiscoveryService;
+import com.hazelcast.util.concurrent.BackoffIdleStrategy;
+import com.hazelcast.util.concurrent.IdleStrategy;
 
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.concurrent.TimeUnit;
 
 import static com.hazelcast.util.Preconditions.checkNotNull;
 import static java.lang.Integer.getInteger;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
 public class DiscoveryJoiner
         extends TcpIpJoiner {
@@ -40,6 +44,9 @@ public class DiscoveryJoiner
 
     private final DiscoveryService discoveryService;
     private final boolean usePublicAddress;
+    private final IdleStrategy idleStrategy =
+            new BackoffIdleStrategy(0, 0, MILLISECONDS.toNanos(10),
+                    MILLISECONDS.toNanos(500));
 
     public DiscoveryJoiner(Node node, DiscoveryService discoveryService, boolean usePublicAddress) {
         super(node);
@@ -49,12 +56,15 @@ public class DiscoveryJoiner
 
     @Override
     protected Collection<Address> getPossibleAddressesForInitialJoin() {
-        Collection<Address> possibleAddresses;
         long deadLine = System.nanoTime() + TimeUnit.SECONDS.toNanos(MAXIMUM_WAITING_FOR_INITIAL_MEMBERS_SECONDS);
-        do {
-            possibleAddresses = getPossibleAddresses();
-        } while (possibleAddresses.isEmpty() && System.nanoTime() < deadLine);
-        return possibleAddresses;
+        for (int i = 0; System.nanoTime() < deadLine; i++) {
+            Collection<Address> possibleAddresses = getPossibleAddresses();
+            if (!possibleAddresses.isEmpty()) {
+                return possibleAddresses;
+            }
+            idleStrategy.idle(i);
+        }
+        return Collections.emptyList();
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/DiscoveryJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/DiscoveryJoiner.java
@@ -28,35 +28,33 @@ import com.hazelcast.util.concurrent.IdleStrategy;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.concurrent.TimeUnit;
 
+import static com.hazelcast.spi.properties.GroupProperty.WAIT_SECONDS_BEFORE_JOIN;
 import static com.hazelcast.util.Preconditions.checkNotNull;
-import static java.lang.Integer.getInteger;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
 
 public class DiscoveryJoiner
         extends TcpIpJoiner {
-    private static final int DEFAULT_MAXIMUM_WAITING_FOR_INITIAL_MEMBERS_SECONDS = 5;
-    private static final String WAITING_FOR_INITIAL_MEMBERS_SECONDS_PROP = "hazelcast.discovery.initial.join.seconds";
-
-    private static final int MAXIMUM_WAITING_FOR_INITIAL_MEMBERS_SECONDS =
-            getInteger(WAITING_FOR_INITIAL_MEMBERS_SECONDS_PROP, DEFAULT_MAXIMUM_WAITING_FOR_INITIAL_MEMBERS_SECONDS);
 
     private final DiscoveryService discoveryService;
     private final boolean usePublicAddress;
     private final IdleStrategy idleStrategy =
             new BackoffIdleStrategy(0, 0, MILLISECONDS.toNanos(10),
                     MILLISECONDS.toNanos(500));
+    private final int maximumWaitingTimeBeforeJoinSeconds;
+
 
     public DiscoveryJoiner(Node node, DiscoveryService discoveryService, boolean usePublicAddress) {
         super(node);
+        this.maximumWaitingTimeBeforeJoinSeconds = node.getProperties().getInteger(WAIT_SECONDS_BEFORE_JOIN);
         this.discoveryService = discoveryService;
         this.usePublicAddress = usePublicAddress;
     }
 
     @Override
     protected Collection<Address> getPossibleAddressesForInitialJoin() {
-        long deadLine = System.nanoTime() + TimeUnit.SECONDS.toNanos(MAXIMUM_WAITING_FOR_INITIAL_MEMBERS_SECONDS);
+        long deadLine = System.nanoTime() + SECONDS.toNanos(maximumWaitingTimeBeforeJoinSeconds);
         for (int i = 0; System.nanoTime() < deadLine; i++) {
             Collection<Address> possibleAddresses = getPossibleAddresses();
             if (!possibleAddresses.isEmpty()) {

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/DiscoveryJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/DiscoveryJoiner.java
@@ -25,11 +25,18 @@ import com.hazelcast.spi.discovery.integration.DiscoveryService;
 
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.concurrent.TimeUnit;
 
 import static com.hazelcast.util.Preconditions.checkNotNull;
+import static java.lang.Integer.getInteger;
 
 public class DiscoveryJoiner
         extends TcpIpJoiner {
+    private static final int DEFAULT_MAXIMUM_WAITING_FOR_INITIAL_MEMBERS_SECONDS = 5;
+    private static final String WAITING_FOR_INITIAL_MEMBERS_SECONDS_PROP = "hazelcast.discovery.initial.join.seconds";
+
+    private static final int MAXIMUM_WAITING_FOR_INITIAL_MEMBERS_SECONDS =
+            getInteger(WAITING_FOR_INITIAL_MEMBERS_SECONDS_PROP, DEFAULT_MAXIMUM_WAITING_FOR_INITIAL_MEMBERS_SECONDS);
 
     private final DiscoveryService discoveryService;
     private final boolean usePublicAddress;
@@ -38,6 +45,16 @@ public class DiscoveryJoiner
         super(node);
         this.discoveryService = discoveryService;
         this.usePublicAddress = usePublicAddress;
+    }
+
+    @Override
+    protected Collection<Address> getPossibleAddressesForInitialJoin() {
+        Collection<Address> possibleAddresses;
+        long deadLine = System.nanoTime() + TimeUnit.SECONDS.toNanos(MAXIMUM_WAITING_FOR_INITIAL_MEMBERS_SECONDS);
+        do {
+            possibleAddresses = getPossibleAddresses();
+        } while (possibleAddresses.isEmpty() && System.nanoTime() < deadLine);
+        return possibleAddresses;
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/spi/discovery/multicast/impl/MulticastDiscoverySender.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/discovery/multicast/impl/MulticastDiscoverySender.java
@@ -67,15 +67,22 @@ public class MulticastDiscoverySender implements Runnable {
     public void run() {
         while (!stop) {
             try {
-                Thread.sleep(SLEEP_DURATION);
-            } catch (InterruptedException e) {
-                logger.finest("Thread sleeping interrupted. This may due to graceful shutdown.");
-            }
-            try {
                 send();
             } catch (IOException e) {
                 logger.finest(e.getMessage());
             }
+            sleepUnlessStopped();
+        }
+    }
+
+    private void sleepUnlessStopped() {
+        if (stop) {
+            return;
+        }
+        try {
+            Thread.sleep(SLEEP_DURATION);
+        } catch (InterruptedException e) {
+            logger.finest("Thread sleeping interrupted. This may due to graceful shutdown.");
         }
     }
 

--- a/hazelcast/src/test/java/com/hazelcast/spi/discovery/DiscoverySpiTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/discovery/DiscoverySpiTest.java
@@ -85,6 +85,7 @@ import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
+import static org.mockito.Mockito.atLeastOnce;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
@@ -339,7 +340,7 @@ public class DiscoverySpiTest extends HazelcastTestSupport {
 
         HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
         try {
-            verify(discoveryService).discoverNodes();
+            verify(discoveryService, atLeastOnce()).discoverNodes();
         } finally {
             instance.getLifecycleService().terminate();
         }

